x/dualstaking/keeper/delegate.go

```diff
@@ -0,0 +1,572 @@
+// commission deduction).
+//
+// The delegated funds are stored in the module's NotBondedPoolName account. Upon
+// request to terminate the delegation, they are moved to the modules BondedPoolName
+// account, and remain locked there for unstakeHoldBlocks witholding period before
+// they are finally released back to the delegator. The timers for bonded funds are
+// tracked are indexed by the delegator, provider, and chainID.
+// The delegation state is stores with fixation using two maps: one for delegations
+// indexed by the combination <provider,chainD,delegator>, used to track delegations
+// and find/access delegations by provider (and chainID); and another for delegators
+// tracking the list of providers for a delegator, indexed by the delegator.
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/lavanet/lava/utils"
+	"github.com/lavanet/lava/utils/slices"
+	"github.com/lavanet/lava/x/dualstaking/types"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	spectypes "github.com/lavanet/lava/x/spec/types"
+	"gopkg.in/yaml.v2"
+)
+
+// getNextEpoch returns the block of next epoch.
+// (all delegate transaction take effect in the subsequent epoch)
+func (k Keeper) getNextEpoch(ctx sdk.Context) (uint64, error) {
+	block := uint64(ctx.BlockHeight())
+	nextEpoch, err := k.epochstorageKeeper.GetNextEpoch(ctx, block)
+	if err != nil {
+		return 0, utils.LavaFormatError("critical: failed to get next epoch", err,
+			utils.Attribute{Key: "block", Value: block},
+		)
+	}
+	return nextEpoch, nil
+}
+
+// validateCoins validates that the input amount is valid and non-negative
+func validateCoins(amount sdk.Coin) error {
+	if !amount.IsValid() {
+		return utils.LavaFormatWarning("invalid coins to delegate",
+			sdkerrors.ErrInvalidCoins,
+			utils.Attribute{Key: "amount", Value: amount},
+		)
+	}
+	return nil
+}
+
+// increaseDelegation increases the delegation of a delegator to a provider for a
+// given chain. It updates the fixation stores for both delegations and delegators,
+// and updates the (epochstorage) stake-entry.
+func (k Keeper) increaseDelegation(ctx sdk.Context, delegator, provider, chainID string, amount sdk.Coin, nextEpoch uint64) error {
+	// get, update and append the delegation entry
+	var delegationEntry types.Delegation
+	prefix := types.DelegationKey(delegator, provider, chainID)
+	found := k.delegationFS.FindEntry(ctx, prefix, nextEpoch, &delegationEntry)
+	if !found {
+		// new delegation (i.e. not increase of existing one)
+		delegationEntry = types.NewDelegation(delegator, provider, chainID)
+	}
+
+	delegationEntry.AddAmount(amount)
+
+	err := k.delegationFS.AppendEntry(ctx, prefix, nextEpoch, &delegationEntry)
+	if err != nil {
+		// append should never fail here
+		return utils.LavaFormatError("critical: append delegation entry", err,
```

Comment:
add the chain ID in the error attributes

Reply:
fixed 96a898ea